File: Android/auikit-service/src/main/java/io/agora/auikit/service/imp/AUIInvitationServiceImpl.kt
package io.agora.auikit.service.imp

import io.agora.auikit.model.AUIInvitationInfo
import io.agora.auikit.model.AUIInvitationInfo.AUIInvitationStatus
import io.agora.auikit.model.AUIUserInfo
import io.agora.auikit.service.IAUIInvitationService
import io.agora.auikit.service.callback.AUICallback
import io.agora.auikit.service.callback.AUIException
import io.agora.auikit.service.collection.AUIAttributesModel
import io.agora.auikit.service.collection.AUICollectionException
import io.agora.auikit.service.collection.AUIListCollection
import io.agora.auikit.service.rtm.AUIRtmManager
import io.agora.auikit.service.rtm.AUIThrottler
import io.agora.auikit.utils.GsonTools
import io.agora.auikit.utils.ObservableHelper
import java.util.Timer
import kotlin.concurrent.scheduleAtFixedRate

const val kInvitationKey = "invitation"

enum class AUIInvitationCmd {
    sendApply,
    cancelApply,
    acceptApply,
    rejectApply,

    sendInvit,
    cancelInvit,
    acceptInvit,
    rejectInvit,
}

class AUIInvitationServiceImpl(
    private val channelName: String,
    private val rtmManager: AUIRtmManager
) : IAUIInvitationService {

    private val invitationCollection = AUIListCollection(channelName, kInvitationKey, rtmManager)

    private val invitationMap = mutableMapOf<String, AUIInvitationInfo>()
    private val checkThrottler by lazy { AUIThrottler() }
    private val observerList = mutableListOf<AUIInvitationInfo>()
    private var timer: Timer? = null
        set(value) {
            value?.cancel()
            field = value
        }

    init {
        invitationCollection.subscribeValueWillChange(this::valueWillChange)
        invitationCollection.subscribeWillAdd(this::metadataWillAdd)
        invitationCollection.subscribeWillMerge(this::metadataWillMerge)
        invitationCollection.subscribeAttributesWillSet(this::metadataWillSet)
        invitationCollection.subscribeAttributesDidChanged(this::onAttributeChanged)
    }


    override fun cleanUserInfo(userId: String, completion: AUICallback?) {
        super.cleanUserInfo(userId, completion)
        if(roomContext.getArbiter(channelName)?.isArbiter() != true){
            return
        }
        invitationCollection.removeMetaData(
            "",
            listOf(
                mapOf("userId" to userId)
            )
        ){
            completion?.onResult(
                if (it == null) null else AUIException(
                    AUIException.ERROR_CODE_RTM_COLLECTION,
                    " $it"
                )
            )
        }
    }

    override fun deInitService(completion: AUICallback?) {
        super.deInitService(completion)
        invitationCollection.cleanMetaData {
            completion?.onResult(
                if (it == null) null else AUIException(
                    AUIException.ERROR_CODE_RTM_COLLECTION,
                    " $it"
                )
            )
        }
        invitationCollection.release()
    }

    private val observableHelper =
        ObservableHelper<IAUIInvitationService.AUIInvitationRespObserver>()

    override fun registerRespObserver(observer: IAUIInvitationService.AUIInvitationRespObserver?) {
        observableHelper.subscribeEvent(observer)
    }

    override fun unRegisterRespObserver(observer: IAUIInvitationService.AUIInvitationRespObserver?) {
        observableHelper.unSubscribeEvent(observer)
    }

    override fun sendInvitation(userId: String, seatIndex: Int, callback: AUICallback?) {
        val info = AUIInvitationInfo()
        info.seatNo = seatIndex
        info.userId = userId
        info.type = AUIInvitationInfo.AUIInvitationType.Invite

        invitationCollection.addMetaData(
            AUIInvitationCmd.sendInvit.name,
            GsonTools.beanToMap(info),
            listOf(
                mapOf("userId" to roomContext.currentUserInfo.userId, "status" to AUIInvitationInfo.AUIInvitationStatus.Waiting),
                mapOf("userId" to roomContext.currentUserInfo.userId, "status" to AUIInvitationInfo.AUIInvitationStatus.Accept),
            )
        ) {
            callback?.onResult(
                if (it == null) null else AUIException(
                    AUIException.ERROR_CODE_RTM_COLLECTION,
                    "$it"
                )
            )
        }
    }

    override fun acceptInvitation(userId: String, seatIndex: Int, callback: AUICallback?) {
        invitationCollection.mergeMetaData(
            AUIInvitationCmd.acceptInvit.name,
            mapOf(
                "status" to AUIInvitationInfo.AUIInvitationStatus.Accept,
                "editTime" to roomContext.ntpTime
            ),
            listOf(mapOf("userId" to userId, "type" to AUIInvitationInfo.AUIInvitationType.Invite))
        ) {
            callback?.onResult(
                if (it == null) null else AUIException(
                    AUIException.ERROR_CODE_RTM_COLLECTION,
                    "$it"
                )
            )
        }
    }

    override fun rejectInvitation(userId: String, callback: AUICallback?) {
        invitationCollection.mergeMetaData(
            AUIInvitationCmd.rejectInvit.name,
            mapOf(
                "status" to AUIInvitationInfo.AUIInvitationStatus.Reject,
                "editTime" to roomContext.ntpTime
            ),
            listOf(mapOf("userId" to userId, "type" to AUIInvitationInfo.AUIInvitationType.Invite))
        ) {
            callback?.onResult(
                if (it == null) null else AUIException(
                    AUIException.ERROR_CODE_RTM_COLLECTION,
                    "$it"
                )
            )
        }
    }

    override fun cancelInvitation(userId: String, callback: AUICallback?) {
        invitationCollection.mergeMetaData(
            AUIInvitationCmd.cancelInvit.name,
            mapOf(
                "status" to AUIInvitationInfo.AUIInvitationStatus.Cancel,
                "editTime" to roomContext.ntpTime
            ),
            listOf(
                mapOf("userId" to userId, "type" to AUIInvitationInfo.AUIInvitationType.Invite)
            )
        ) {
            callback?.onResult(
                if (it == null) null else AUIException(
                    AUIException.ERROR_CODE_RTM_COLLECTION,
                    "$it"
                )
            )
        }
    }

    override fun sendApply(seatIndex: Int, callback: AUICallback) {
        val info = AUIInvitationInfo()
        info.seatNo = seatIndex
        info.userId = roomContext.currentUserInfo.userId
        info.type = AUIInvitationInfo.AUIInvitationType.Apply

        invitationCollection.addMetaData(
            AUIInvitationCmd.sendApply.name,
            GsonTools.beanToMap(info),
            listOf(
                mapOf("userId" to info.userId, "status" to AUIInvitationStatus.Waiting),
                mapOf("userId" to info.userId, "status" to AUIInvitationStatus.Accept)
            )
        ) {
            callback.onResult(
                if (it == null) null else AUIException(
                    AUIException.ERROR_CODE_RTM_COLLECTION,
                    "$it"
                )
            )
        }
    }

    override fun cancelApply(callback: AUICallback) {
        cancelApplyByUserId(roomContext.currentUserInfo.userId, callback)
    }

    private fun cancelApplyByUserId(userId: String, callback: AUICallback) {
        invitationCollection.mergeMetaData(
            AUIInvitationCmd.cancelApply.name,
            mapOf(
                "status" to AUIInvitationStatus.Cancel,
                "editTime" to roomContext.ntpTime
            ),
            listOf(
                mapOf(
                    "userId" to userId,
                    "type" to AUIInvitationInfo.AUIInvitationType.Apply
                )
            )
        ) {
            callback.onResult(
                if (it == null) null else AUIException(
                    AUIException.ERROR_CODE_RTM_COLLECTION,
                    "$it"
                )
            )
        }
    }

    override fun acceptApply(userId: String, seatIndex: Int, callback: AUICallback) {
        invitationCollection.mergeMetaData(
            AUIInvitationCmd.acceptApply.name,
            mapOf(
                "status" to AUIInvitationStatus.Accept,
                "editTime" to roomContext.ntpTime
            ),
            listOf(
                mapOf("userId" to userId, "type" to AUIInvitationInfo.AUIInvitationType.Apply)
            )
        ) {
            callback.onResult(
                if (it == null) null else AUIException(
                    AUIException.ERROR_CODE_RTM_COLLECTION,
                    "$it"
                )
            )
        }
    }

    override fun rejectApply(userId: String, callback: AUICallback) {
        invitationCollection.mergeMetaData(
            AUIInvitationCmd.rejectApply.name,
            mapOf(
                "status" to AUIInvitationInfo.AUIInvitationStatus.Reject,
                "editTime" to roomContext.ntpTime
            ),
            listOf(mapOf("userId" to userId, "type" to AUIInvitationInfo.AUIInvitationType.Apply))
        ) {
            callback.onResult(
                if (it == null) null else AUIException(
                    AUIException.ERROR_CODE_RTM_COLLECTION,
                    "$it"
                )
            )
        }
    }


    override fun getChannelName() = channelName

    private fun valueWillChange(
        publisherId: String,
        valueCmd: String?,
        value: Map<String, Any>
    ): Map<String, Any>? {
        val tempItem = mutableMapOf<String, Any>()
        tempItem.putAll(value)
        val currentTime = roomContext.ntpTime
        tempItem["editTime"] = currentTime
        when (valueCmd) {
            AUIInvitationCmd.sendInvit.name,
            AUIInvitationCmd.sendApply.name -> tempItem["createTime"] = currentTime
        }
        return tempItem
    }

    private fun metadataWillAdd(
        publisherId: String,
        valueCmd: String?,
        value: Map<String, Any>,
        attrs: AUIAttributesModel
    ): AUICollectionException? {
        return when (valueCmd) {
            AUIInvitationCmd.sendApply.name, AUIInvitationCmd.sendInvit.name -> {
                attrs.setList(attrs.getList()?.filter { it["userId"] != value["userId"] })
                null
            }
            else -> {
                AUICollectionException.ErrorCode.unknown.toException("UnKnown cmd : $valueCmd")
            }
        }
    }

    private fun metadataWillMerge(
        publisherId: String,
        valueCmd: String?,
        newValue: Map<String, Any>,
        oldValue: Map<String, Any>
    ): AUICollectionException? {
        return when (valueCmd) {
            AUIInvitationCmd.acceptInvit.name -> {
                val userId = oldValue["userId"] as? String ?: ""
                val seatIndex = (oldValue["seatNo"] as? Long)?.toInt() ?: 0
                var error: AUIException? = null
                observableHelper.notifyEventHandlers {
                    error = it.onInviteWillAccept(userId, seatIndex)
                }
                if (error != null) AUICollectionException.ErrorCode.unknown.toException() else null
            }

            AUIInvitationCmd.acceptApply.name -> {
                val userId = oldValue["userId"] as? String ?: ""
                val seatIndex = (oldValue["seatNo"] as? Long)?.toInt() ?: 0
                var error: AUIException? = null
                observableHelper.notifyEventHandlers {
                    error = it.onApplyWillAccept(userId, seatIndex)
                }
                if (error != null) AUICollectionException.ErrorCode.unknown.toException() else null
            }

            AUIInvitationCmd.cancelInvit.name,
            AUIInvitationCmd.cancelApply.name,
            AUIInvitationCmd.rejectInvit.name,
            AUIInvitationCmd.rejectApply.name -> null
            else -> AUICollectionException.ErrorCode.unknown.toException()
        }
    }

    private fun metadataWillSet(
        channelName: String,
        observeKey: String,
        valueCmd: String?,
        value: AUIAttributesModel
    ) {
        val list = value.getList() ?: return
        val currentTime = roomContext.ntpTime
        val filterList = list.filter { attr ->
            val editTime = attr["editTime"] as? Long ?: 0
            val invalidTs = attr["invalidTs"] as? Long ?: 0
            val status = attr["status"] as? Int ?: 0
            if (status == AUIInvitationInfo.AUIInvitationStatus.Waiting) {
                return@filter true
            }
            if (currentTime - editTime < invalidTs) {
                return@filter true
            }
            return@filter false
        }
        value.setList(filterList)
    }

    private fun onAttributeChanged(channelName: String, key: String, value: AUIAttributesModel) {
        val list =
            GsonTools.toList(GsonTools.beanToString(value.getList()), AUIInvitationInfo::class.java)
                ?: return

        // 申请列表
        val newApplyList = list.filter { it.type == AUIInvitationInfo.AUIInvitationType.Apply }
        val oldApplyList =
            invitationMap.values.filter { it.type == AUIInvitationInfo.AUIInvitationType.Apply }

        // 邀请列表
        val newInviteList = list.filter { it.type == AUIInvitationInfo.AUIInvitationType.Invite }
        val oldInviteList =
            invitationMap.values.filter { it.type == AUIInvitationInfo.AUIInvitationType.Invite }

        invitationMap.clear()
        list.forEach {
            invitationMap[it.userId] = it
        }

        // 处理申请差异列表
        var applyListChanged = false
        newApplyList.forEach { newApply ->
            val oldApply = oldApplyList.find { it.userId == newApply.userId }
            if (oldApply?.status != newApply.status) {
                applyListChanged = true
                when (newApply.status) {
                    AUIInvitationInfo.AUIInvitationStatus.Accept -> {
                        observableHelper.notifyEventHandlers {
                            it.onApplyAccepted(newApply.userId, newApply.seatNo)
                        }
                    }

                    AUIInvitationInfo.AUIInvitationStatus.Reject -> {
                        observableHelper.notifyEventHandlers {
                            it.onApplyRejected(newApply.userId)
                        }
                    }

                    AUIInvitationInfo.AUIInvitationStatus.Timeout -> {
                    }

                    AUIInvitationInfo.AUIInvitationStatus.Waiting -> {
                        observableHelper.notifyEventHandlers {
                            it.onReceiveNewApply(newApply.userId, newApply.seatNo)
                        }
                    }
                }
            }
        }
        oldApplyList.forEach { oldApply ->
            val newApply = newApplyList.find { it.userId == oldApply.userId }
            if (newApply == null) {
                if (oldApply.status == AUIInvitationInfo.AUIInvitationStatus.Waiting) {
                    observableHelper.notifyEventHandlers {
                        it.onApplyCanceled(oldApply.userId)
                    }
                }
                applyListChanged = true
                return@forEach
            }
        }
        if (applyListChanged) {
            observableHelper.notifyEventHandlers {
                it.onApplyListUpdate(newApplyList.filter { it.status == AUIInvitationStatus.Waiting }.map {
                    AUIUserInfo().apply {
                        userId = it.userId
                        micIndex = it.seatNo
                    }
                })
            }
        }

        // 处理邀请差异列表
        newInviteList.forEach { newInvite ->
            val oldInvite = oldApplyList.find { it.userId == newInvite.userId }
            if (oldInvite?.status != newInvite.status) {
                when (newInvite.status) {
                    AUIInvitationInfo.AUIInvitationStatus.Accept -> {
                        observableHelper.notifyEventHandlers {
                            it.onInviteeAccepted(newInvite.userId, newInvite.seatNo)
                        }
                    }

                    AUIInvitationInfo.AUIInvitationStatus.Reject -> {
                        observableHelper.notifyEventHandlers {
                            it.onInviteeRejected(newInvite.userId)
                        }
                    }

                    AUIInvitationInfo.AUIInvitationStatus.Timeout -> {

                    }

                    AUIInvitationInfo.AUIInvitationStatus.Waiting -> {
                        observableHelper.notifyEventHandlers {
                            it.onReceiveInvitation(newInvite.userId, newInvite.seatNo)
                        }
                    }
                }
            }
        }
        oldInviteList.forEach { oldInvite ->
            val newInvite = newApplyList.find { it.userId == oldInvite.userId }
            if (newInvite == null && oldInvite.status == AUIInvitationInfo.AUIInvitationStatus.Waiting) {
                observableHelper.notifyEventHandlers {
                    it.onInvitationCancelled(oldInvite.userId)
                }
                return@forEach
            }
        }

        checkThrottler.triggerLastEvent(delay = 300) {
            checkWaitingTimeout()
        }
    }

    private fun checkWaitingTimeout() {
        if (roomContext.getArbiter(channelName)?.isArbiter() != true) {
            return
        }
        val observerList =
            invitationMap.values.filter { it.status == AUIInvitationInfo.AUIInvitationStatus.Waiting }
        if (observerList.isEmpty()) {
            timer = null
            return
        }
        this.observerList.clear()
        this.observerList.addAll(observerList)
        timer = Timer().apply {
            scheduleAtFixedRate(0, 1000) {
                val currentTs = roomContext.ntpTime
                val observerList = ArrayList(this@AUIInvitationServiceImpl.observerList)
                this@AUIInvitationServiceImpl.observerList.clear()
                observerList.forEach { info ->
                    if (currentTs - info.createTime > info.timeoutTs) {
                        when (info.type) {
                            AUIInvitationInfo.AUIInvitationType.Apply -> cancelApplyByUserId(info.userId) {}
                            AUIInvitationInfo.AUIInvitationType.Invite -> cancelInvitation(info.userId) {}
                        }
                    } else {
                        this@AUIInvitationServiceImpl.observerList.add(info)
                    }
                }
                if (this@AUIInvitationServiceImpl.observerList.isEmpty()) {
                    timer = null
                }
            }
        }
    }
}

File: iOS/AUIKitCore/Sources/Service/Impl/AUIInvitationServiceImpl.swift
//
//  AUIInvitationServiceImpl.swift
//  AUIKit
//
//  Created by wushengtao on 2023/2/23.
//

import Foundation
import AgoraRtcKit
import YYModel

fileprivate let AUIInvitationKey = "invitation"

private enum AUIInvitationCmd: String {
    case sendApply = "sendApply"
    case cancelApply = "cancelApply"
    case acceptApply = "acceptApply"
    case rejectApply = "rejectApply"
    
    case sendInvit = "sendInvit"
    case cancelInvit = "cancelInvit"
    case acceptInvit = "acceptInvit"
    case rejectInvit = "rejectInvit"
}

private let kEditTimeKey = "editTime"
private let kStatusKey = "status"
private let kUserIdKey = "userId"
private let kTypeKey = "type"
private let kSeatNoKey = "seatNo"

//邀请Service实现
@objc open class AUIInvitationServiceImpl: NSObject {
    private var respDelegates: NSHashTable<AUIInvitationRespDelegate> = NSHashTable<AUIInvitationRespDelegate>.weakObjects()
    private var channelName: String!
    private var rtmManager: AUIRtmManager!
    private var invitationCollection:AUIListCollection!
    private var invitationMap: [String: AUIInvitationInfo] = [:]
    
    lazy var checkThrottler: AUIThrottler = AUIThrottler()
    private var observerList: [AUIInvitationInfo] = []
    private var timer: Timer? {
        didSet {
            oldValue?.invalidate()
        }
    }
    
    deinit {
        aui_info("deinit AUIInvitationServiceImpl", tag: "AUIInvitationServiceImpl")
    }
    
    public init(channelName: String, rtmManager: AUIRtmManager) {
        self.channelName = channelName
        self.rtmManager = rtmManager
        self.invitationCollection = AUIListCollection(channelName: channelName,
                                                      observeKey: AUIInvitationKey,
                                                      rtmManager: rtmManager)
        super.init()
        invitationCollection.subsceibeValueWillChange {[weak self] publisherId, dataCmd, newItem in
            return self?.valueWillChange(publiserId: publisherId,
                                         dataCmd: dataCmd,
                                         newItem: newItem)
        }
        invitationCollection.subscribeWillAdd {[weak self] publisherId, dataCmd, newItem, attr in
            return self?.metadataWillAdd(publiserId: publisherId,
                                         dataCmd: dataCmd,
                                         newItem: newItem,
                                         attr: attr)
        }
        
        invitationCollection.subscribeWillMerge {[weak self] publisherId, dataCmd, updateMap, currentMap in
            return self?.metadataWillMerge(publiserId: publisherId, dataCmd: dataCmd, updateMap: updateMap, currentMap: currentMap)
        }
        
        invitationCollection.subscribeAttributesWillSet {[weak self] channelName, key, valueCmd, attr in
            guard let self = self else { return }
            guard let value = attr.getList() else { return }
            let currentTime = self.getRoomContext().getNtpTime()
            //把已经accept的麦位申请&邀请给reject
            var filterList = self.rejectDuplicateAcceptedRequests(list: value)
            filterList = filterList.filter { attr in
                let editTime = attr[kEditTimeKey] as? Int64 ?? 0
                let invalidTs = attr["invalidTs"] as? Int64 ?? kInvitationInvalidTs
                let status = attr[kStatusKey] as? Int
                if status == AUIInvitationStatus.waiting.rawValue {
                    return true
                }
                //过滤超时的无效(非waitting)数据
                if currentTime - editTime < invalidTs { return true }
                return false
            }
            aui_info("invitation will set filter list: \(filterList.count) / \(value.count)", tag: "AUIInvitationServiceImpl")
            attr.setList(filterList)
        }
        
        invitationCollection.subscribeAttributesDidChanged {[weak self] channelName, key, value in
            self?.onAttributesDidChanged(channelName: channelName, key: key, value: value)
        }

        aui_info("init AUIInvitationServiceImpl", tag: "AUIInvitationServiceImpl")
    }
}

extension AUIInvitationServiceImpl {
    //检查是否需要开启定时器：1.邀请申请列表变更， 2.仲裁者切换
    private func checkWaittingTimeout() {
        guard getRoomContext().getArbiter(channelName: channelName)?.isArbiter() ?? false else { return }
        let observerList = invitationMap.compactMap({$1.status == .waiting ? $1 : nil})
        if observerList.isEmpty {
            timer = nil
            return
        }
        self.observerList = observerList
        timer = Timer.scheduledTimer(withTimeInterval: 1, repeats: true, block: {[weak self] t in
            guard let self = self else {return}
            let currentTs = self.getRoomContext().getNtpTime()
            let observerList = self.observerList
            self.observerList.removeAll()
            observerList.forEach { info in
                if currentTs - info.createTime > info.timeoutTs {
                    if info.type == .apply {
                        aui_info("apply checkWaittingTimeout: \(info.userId)", tag: "AUIInvitationServiceImpl")
                        self.cancelApply { err in }
                    } else {
                        aui_info("invit checkWaittingTimeout: \(info.userId)", tag: "AUIInvitationServiceImpl")
                        self.cancelInvitation(userId: info.userId) { err in }
                    }
                } else {
                    self.observerList.append(info)
                }
            }
            if self.observerList.isEmpty {
                self.timer = nil
            }
        })
        timer?.fire()
    }
    
    private func rejectDuplicateAcceptedRequests(list: [[String: Any]]) -> [[String: Any]] {
        //获取accept的麦位数组
        let acceptSeatNos = list.map({ item in
            if item[kStatusKey] as? Int == AUIInvitationStatus.accept.rawValue {
                return item[kSeatNoKey] as? Int
            }
            return nil
        })
        
        var updateList = [[String: Any]]()
        for var item in list {
            //如果accept过的麦位还有申请的，设置为reject
            if let seatNo = item[kSeatNoKey] as? Int,
               let status = item[kStatusKey] as? Int,
               status == AUIInvitationStatus.waiting.rawValue,
                acceptSeatNos.contains(seatNo) {
                let currentTime = self.getRoomContext().getNtpTime()
                item[kStatusKey] = AUIInvitationStatus.reject.rawValue
                item[kEditTimeKey] = currentTime
            }
            updateList.append(item)
        }
        
        return updateList
    }
    
    private func valueWillChange(publiserId: String,
                                 dataCmd: String?,
                                 newItem: [String: Any]) -> [String: Any]? {
        guard let dataCmd = AUIInvitationCmd(rawValue: dataCmd ?? "") else {
            return newItem
        }
        var tempItem = newItem
        let currentTime = self.getRoomContext().getNtpTime()
        tempItem[kEditTimeKey] = currentTime
        switch dataCmd {
        case .sendApply, .sendInvit:
            tempItem["createTime"] = currentTime
            break
        default:
            break
        }
        
        return tempItem
    }
    
    private func metadataWillAdd(publiserId: String,
                                 dataCmd: String?,
                                 newItem: [String: Any],
                                 attr: AUIAttributesModel) -> NSError? {
        guard let dataCmd = AUIInvitationCmd(rawValue: dataCmd ?? "") else {
            return AUICommonError.unknown.toNSError()
        }
        
        guard let model = AUIInvitationInfo.yy_model(with: newItem) else {
            return AUICommonError.unknown.toNSError()
        }
        switch dataCmd {
        case .sendApply, .sendInvit:
            //能走到这里，如果有老的userId的数据，需要清理掉，防止重复，因为add请求的时候过滤了status为waitting/accept的，剩余状态的无效数据需要清理
            if var list = attr.getList() {
                list = list.filter { $0[kUserIdKey] as? String != model.userId }
                attr.setList(list)
            }
            return nil
        default:
            break
        }
        
        return NSError.auiError("add invitation cmd incorrect")
    }
    
    private func metadataWillMerge(publiserId: String,
                                   dataCmd: String?,
                                   updateMap: [String: Any],
                                   currentMap: [String: Any]) -> NSError? {
        guard let dataCmd = AUIInvitationCmd(rawValue: dataCmd ?? "") else {
            return AUICommonError.unknown.toNSError()
        }
        
        let userId: String = currentMap[kUserIdKey] as? String ?? ""
        let seatIndex: Int = currentMap[kSeatNoKey] as? Int ?? 0
        let metaData = NSMutableDictionary()
        var err: NSError? = nil
        switch dataCmd {
        case .acceptApply:
            for obj in respDelegates.allObjects {
                err = obj.onApplyWillAccept?(userId: userId, seatIndex: seatIndex, metaData: metaData)
            }
        case .acceptInvit:
            for obj in respDelegates.allObjects {
                err = obj.onInviteWillAccept?(userId: userId, seatIndex: seatIndex, metaData: metaData)
            }
        default:
            break
        }
        
        return err
    }
    
    private func onAttributesDidChanged(channelName: String, key: String, value: AUIAttributesModel) {
        guard AUIInvitationKey == key else { return}
        aui_info("recv chorus attr did changed \(value)", tag: "AUIInvitationServiceImpl")
        guard let invitationArray = value.getList(),
              let invitationList = NSArray.yy_modelArray(with: AUIInvitationInfo.self, json: invitationArray) as? [AUIInvitationInfo] else {
            return
        }
        
        let oldMap = self.invitationMap
        let newMap = Dictionary(uniqueKeysWithValues: invitationList.map { ($0.userId, $0) })
        self.invitationMap = newMap
        var applyList: [AUIInvitationInfo] = []
        var invitList: [AUIInvitationInfo] = []
        newMap.forEach { userId, newInfo in
            let oldInfo = oldMap[userId]
            let isTargetUser = newInfo.userId == getRoomContext().currentUserInfo.userId
            if oldInfo?.type == newInfo.type || oldInfo == nil {
                if newInfo.type == .invite {
                    if newInfo.status == .waiting {
                        invitList.append(newInfo)
                    }
                    if oldInfo?.status != newInfo.status {
                        if isTargetUser {
                            switch newInfo.status {
                            case .waiting:
                                self.respDelegates.allObjects.forEach {
                                    $0.onReceiveNewInvitation(userId: userId, seatIndex: newInfo.seatNo)
                                }
                            case .cancel, .timeout:
                                self.respDelegates.allObjects.forEach {
                                    $0.onInvitationCancelled(userId: userId)
                                }
                            case .reject:
                                self.respDelegates.allObjects.forEach {
                                    $0.onInviteeRejected(userId: userId)
                                }
                            case .accept:
                                self.respDelegates.allObjects.forEach {
                                    $0.onInviteeAccepted(userId: userId)
                                }
                            }
                        }
                    }
                } else if newInfo.type == .apply {
                    if newInfo.status == .waiting {
                        applyList.append(newInfo)
                    }
                    if oldInfo?.status != newInfo.status {
                        if isTargetUser {
                            switch newInfo.status {
                            case .waiting:
                                self.respDelegates.allObjects.forEach {
                                    $0.onReceiveNewApply(userId: userId, seatIndex: newInfo.seatNo)
                                }
                            case .cancel, .timeout:
                                self.respDelegates.allObjects.forEach {
                                    $0.onApplyCanceled(userId: userId)
                                }
                            case .reject:
                                self.respDelegates.allObjects.forEach {
                                    $0.onApplyRejected(userId: userId)
                                }
                            case .accept:
                                self.respDelegates.allObjects.forEach {
                                    $0.onApplyAccepted(userId: userId)
                                }
                            }
                        }
                    }
                }
            } else {
                aui_warn("invitation type changed \(oldInfo?.type.rawValue ?? -1) -> \(newInfo.type.rawValue)", tag: "AUIInvitationServiceImpl")
            }
        }
        
        self.respDelegates.allObjects.forEach {
            $0.onReceiveApplyUsersUpdate(applyList: applyList)
        }
        self.respDelegates.allObjects.forEach {
            $0.onInviteeListUpdate(inviteeList: invitList)
        }
        
        checkThrottler.triggerLastEvent(after: 0.3) {
            self.checkWaittingTimeout()
        }
    }
}

extension AUIInvitationServiceImpl: AUIInvitationServiceDelegate {
    
    public func getRoomContext() -> AUIRoomContext {
        return AUIRoomContext.shared
    }
    
    public func getChannelName() -> String {
        return channelName
    }
    
    public func bindRespDelegate(delegate: AUIInvitationRespDelegate) {
        respDelegates.add(delegate)
    }
    
    public func unbindRespDelegate(delegate: AUIInvitationRespDelegate) {
        respDelegates.remove(delegate)
    }
    
    public func sendInvitation(userId: String, seatIndex: Int?, callback: @escaping (NSError?) -> ()) {
        let model = AUIInvitationInfo()
        model.type = .invite
        model.seatNo = seatIndex ?? 0
        model.userId = userId
        guard let value = model.yy_modelToJSONObject() as? [String: Any] else {
            callback(NSError.auiError("convert to json fail"))
            return
        }
        //当当前状态是waitting和accept时不允许写入，无论邀请申请
        let filter = [
            [kUserIdKey: model.userId, kStatusKey: AUIInvitationStatus.waiting.rawValue],
            [kUserIdKey: model.userId, kStatusKey: AUIInvitationStatus.accept.rawValue]
        ]
        invitationCollection.addMetaData(valueCmd: AUIInvitationCmd.sendInvit.rawValue,
                                         value: value, 
                                         filter: filter,
                                         callback: callback)
    }
    
    public func acceptInvitation(userId: String, seatIndex: Int?, callback: @escaping (NSError?) -> ()) {
        let value: [String: Any] = [kStatusKey: AUIInvitationStatus.accept.rawValue]
        invitationCollection.mergeMetaData(valueCmd: AUIInvitationCmd.acceptInvit.rawValue,
                                           value: value,
                                           filter: [[kUserIdKey: userId, kTypeKey: AUIInvitationType.invite.rawValue]],
                                           callback: callback)
    }
    
    public func rejectInvitation(userId: String, callback: @escaping (NSError?) -> ()) {
        let value: [String: Any] = [kStatusKey: AUIInvitationStatus.reject.rawValue]
        invitationCollection.mergeMetaData(valueCmd: AUIInvitationCmd.rejectInvit.rawValue,
                                           value: value,
                                           filter: [[kUserIdKey: userId, kTypeKey: AUIInvitationType.invite.rawValue]],
                                           callback: callback)
    }
    
    public func cancelInvitation(userId: String, callback: @escaping (NSError?) -> ()) {
        let value: [String: Any] = [kStatusKey: AUIInvitationStatus.cancel.rawValue]
        invitationCollection.mergeMetaData(valueCmd: AUIInvitationCmd.cancelInvit.rawValue,
                                           value: value,
                                           filter: [[kUserIdKey: userId, kTypeKey: AUIInvitationType.invite.rawValue]],
                                           callback: callback)
    }
    
    public func sendApply(seatIndex: Int?, callback: @escaping (NSError?) -> ()) {
        let model = AUIInvitationInfo()
        model.type = .apply
        model.seatNo = seatIndex ?? 0
        model.userId = getRoomContext().currentUserInfo.userId
        guard let value = model.yy_modelToJSONObject() as? [String: Any] else {
            callback(NSError.auiError("convert to json fail"))
            return
        }
        //当当前状态是waitting和accept时不允许写入，无论邀请申请
        let filter = [
            [kUserIdKey: model.userId, kStatusKey: AUIInvitationStatus.waiting.rawValue],
            [kUserIdKey: model.userId, kStatusKey: AUIInvitationStatus.accept.rawValue]
        ]
        invitationCollection.addMetaData(valueCmd: AUIInvitationCmd.sendApply.rawValue,
                                         value: value,
                                         filter: filter,
                                         callback: callback)
    }
    
    public func cancelApply(callback: @escaping (NSError?) -> ()) {
        let value: [String: Any] = [kStatusKey: AUIInvitationStatus.cancel.rawValue]
        let userId = getRoomContext().currentUserInfo.userId
        invitationCollection.mergeMetaData(valueCmd: AUIInvitationCmd.cancelApply.rawValue,
                                           value: value,
                                           filter: [[kUserIdKey: userId, kTypeKey: AUIInvitationType.apply.rawValue]],
                                           callback: callback)
    }
    
    public func acceptApply(userId: String, seatIndex: Int?, callback: @escaping (NSError?) -> ()) {
        let value: [String: Any] = [kStatusKey: AUIInvitationStatus.accept.rawValue]
        
        invitationCollection.mergeMetaData(valueCmd: AUIInvitationCmd.acceptApply.rawValue,
                                           value: value,
                                           filter: [[kUserIdKey: userId, kTypeKey: AUIInvitationType.apply.rawValue]],
                                           callback: callback)
    }
    
    public func rejectApply(userId: String, callback: @escaping (NSError?) -> ()) {
        let value: [String: Any] = [kStatusKey: AUIInvitationStatus.reject.rawValue]
        invitationCollection.mergeMetaData(valueCmd: AUIInvitationCmd.rejectApply.rawValue,
                                           value: value,
                                           filter: [[kUserIdKey: userId, kTypeKey: AUIInvitationType.apply.rawValue]],
                                           callback: callback)
    }
}

extension AUIInvitationServiceImpl {
    public func initService(completion: @escaping ((NSError?) -> ())){
    }
    
    public func cleanUserInfo(userId: String, completion: @escaping ((NSError?) -> ())) {
        invitationCollection.removeMetaData(valueCmd: nil,
                                            filter: [[kUserIdKey: userId]],
                                            callback: completion)
    }
    
    public func deinitService(completion: @escaping ((NSError?) -> ())) {
        invitationCollection.cleanMetaData(callback: completion)
    }
}
